### ISSUE TITLE
[`RUF`] `RUF056` false negative on shadowed bindings

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ruff#18855]
### Summary

Like #18777, [falsy-dict-get-fallback (RUF056)](https://docs.astral.sh/ruff/rules/falsy-dict-get-fallback/#falsy-dict-get-fallback-ruf056) has a false negative when the name is a shadowed binding. [playground](https://play.ruff.rs/393aed06-5ef5-420d-9152-12f92affe66a)
```py
dict = {}
if dict.get(key, False):  # No error
    ...
```

### Version

playground

**Comment by MichaReiser** [comment]
Using `resolve_name` over `only_binding` can lead to new false positives when control flow is involved:

```py
def demonstrate_false_positive():
    if True:
        a= set()
    else:
        a = {}


    # dict is now a string, should NOT trigger RUF056
    if a.get(key, False):  # But this DOES trigger - FALSE POSITIVE
        pass
```

That's why the fix is more complicated than replacing one with the other. I think the proper fix is to change `only_binding` to return `Some` if the last visible binding is visible regardless of the control flow graph taken. This would still leave some false positives (e.g the example above) but it would be an improvement over what we have today. 

The most comprehensive solution would be to enhance our type inference to verify that all potentially visible bindings resolve to a dictionary.
